### PR TITLE
E2E (integration) tests for `HSCS`

### PR DESCRIPTION
**Pull request by joan41868**
Signed-off-by: Yoan Sredkov <yoansredkov@gmail.com>

### CODE REVIEW

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    @Override
+    protected List<HapiApiSpec> getSpecsInSuite() {
+        return List.of(
+//                HSCS_EVM_008_SelfDesctructUpdatesHederaAccount(),
+                HSCS_EVM_0014_ContractCreateFailsOnLargeContracts()
+        );
+    }
+
+    // Done
+    private HapiApiSpec HSCS_EVM_008_SelfDesctructUpdatesHederaAccount() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:50]: I think that it would be better to place the selfdestruct test in the `opcodes` package -> `SelfDestructSuite.java`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                                .hasCostAnswerPrecheck(ACCOUNT_DELETED),
+                        getContractInfo("selfDestroying")
+                                .hasCostAnswerPrecheck(CONTRACT_DELETED)
+                );
+    }
+
+    // Validates that ContractCreate fails to create a contract > 24KiB
+    // By uploading a LARGE file with multiple file appends, it is possible
+    // However, trying to do it at 1 tx will result in TRANSACTION_OVERSIZE
+    private HapiApiSpec HSCS_EVM_0014_ContractCreateFailsOnLargeContracts() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:73]: I think that it would be better to place this test in `ContractCreateSuite.java`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private HapiApiSpec HSCS_EVM_005_TransferOfHBarsWorksBetweenContracts() {
+        return defaultHapiSpec("HSCS_EVM_005_TransferOfHBarsWorksBetweenContracts")
+                .given()
+                .when()
+                .then();
+    }
+
+    private HapiApiSpec HSCS_EVM_004_GasRefundingLogicWorksAsExpected() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:124]: I think that we should place this logic in the SSTORESuite in `opcode` package

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private HapiApiSpec HSCS_EVM_006_ContractHBarTransferToAccount() {
+        return defaultHapiSpec("HSCS_EVM_006_ContractHBarTransferToAccount")
+                .given()
+                .when()
+                .then();
+    }
+
+    private HapiApiSpec HSCS_EVM_005_TransferOfHBarsWorksBetweenContracts() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:117]: We can place this in the `ContractCallSuite`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private HapiApiSpec HSCS_EVM_0010_MultiSignatureAccounts() {
+        return defaultHapiSpec("HSCS_EVM_009_MinimumChargeOfTxnGasLimit")
+                .given()
+                .when()
+                .then();
+    }
+
+    private HapiApiSpec HSCS_EVM_006_ContractHBarTransferToAccount() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:110]: We can place this in the `ContractCallSuite`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                                .bytecode("bytecode").via("cc")
+                                .payingWith(PAYER)
+                                .hasPrecheck(OK).hasAnyKnownStatus(),
+
+                        getTxnRecord("cc").logged()
+                )
+                .then();
+    }
+
+    private HapiApiSpec HSCS_EVM_0010_MultiSignatureAccounts() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:103]: We can place this in the `ContractCallSuite`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                );
+    }
+
+    // Validates that ContractCreate fails to create a contract > 24KiB
+    // By uploading a LARGE file with multiple file appends, it is possible
+    // However, trying to do it at 1 tx will result in TRANSACTION_OVERSIZE
+    private HapiApiSpec HSCS_EVM_0014_ContractCreateFailsOnLargeContracts() {
+        final var PAYER = "payer";
+        final var data = ERC1155ContractInteractions.getFileContents(ContractResources.LARGE_CONTRACT);
+        return defaultHapiSpec("HSCS_EVM_0014_ContractCreateFailsOnLargeContracts")
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:76]: I think that we can use the `updateLargeFile` method

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/openzeppelin/ERC1155ContractInteractions.java [hunk 8]
```diff
@@ -57,7 +57,7 @@ public static void main(String... args) {
 		new ERC1155ContractInteractions().runSuiteSync();
 	}
 	
-	private ByteString getFileContents(String path) {
+	public static ByteString getFileContents(String path) {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/openzeppelin/ERC1155ContractInteractions.java:60]: unnecessary change

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java (hunk 7 above)
Comment (joan41868) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:83]: I thought so as well, but I got `INVALID_SIGNATURE` when I used it. Will investigate further.

File: test-clients/src/main/resource/contract/bytecodes/LargeContract.bin [hunk 9]
```diff
@@ -0,0 +1 @@
...6e73756666696369656e742062616c616e636520666f60008201527f72207472616e7366657200000000000000000000000000000000000000000000602082015250565b7f455243313135353a2073657474696e6720617070726f76616c2073746174757360008201527f20666f722073656c660000000000000000000000000000000000000000000000602082015250565b7f455243313135353a206163636f756e747320616e6420696473206c656e67746860008201527f206d69736d617463680000000000000000000000000000000000000000000000602082015250565b7f455243313135353a2069647320616e6420616d6f756e7473206c656e6774682060008201527f6d69736d61746368000000000000000000000000000000000000000000000000602082015250565b600060443d101561249457612517565b61249c611d9d565b60043d036004823e80513d602482011167ffffffffffffffff821117156124c4575050612517565b808201805167ffffffffffffffff8111156124e25750505050612517565b80602083010160043d0385018111156124ff575050505050612517565b61250e82602001850186611fdf565b82955050505050505b90565b61252381611ef7565b811461252e57600080fd5b50565b61253a81611f09565b811461254557600080fd5b50565b61255181611f15565b811461255c57600080fd5b50565b61256881611f61565b811461257357600080fd5b5056fea2646970667358221220af581ac064a50b92c5abaf75f02c2d92d0547d6e0432e5f94b1af42bed54870a64736f6c63430008070033
```
Comment (shemnon) [inline test-clients/src/main/resource/contract/bytecodes/LargeContract.bin:1]: Where is the solidity source for this?

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+import static com.hedera.services.bdd.spec.transactions.TxnVerbs.contractCall;
+import static com.hedera.services.bdd.spec.transactions.TxnVerbs.contractCreate;
+import static com.hedera.services.bdd.spec.transactions.TxnVerbs.cryptoCreate;
+import static com.hedera.services.bdd.spec.transactions.TxnVerbs.fileCreate;
+import static com.hedera.services.bdd.spec.utilops.CustomSpecAssert.allRunFor;
+import static com.hedera.services.bdd.spec.utilops.UtilVerbs.assertionsHold;
+import static com.hedera.services.bdd.spec.utilops.UtilVerbs.withOpContext;
+
+/**
+ * Major todo - find how to use HBars in smart contracts
```
Comment (joan41868) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/integration/EvmSuites.java:25]: will remove

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/ContractCreateSuite.java [hunk 11]
```diff
@@ -83,9 +92,52 @@ public static void main(String... args) {
 						rejectsInsufficientGas(),
 						createsVanillaContractAsExpectedWithOmittedAdminKey(),
 						childCreationsHaveExpectedKeysWithOmittedAdminKey(),
+						cannotCreateTooLargeContract()
 				}
 		);
 	}
+	/* Attempts to create a 46 KB contract */
```
Comment (joan41868) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/ContractCreateSuite.java:99]: @Daniel-K-Ivanov

File: test-clients/src/main/resource/contract/bytecodes/LargeContract.bin (hunk 9 above)
Comment (joan41868) [reply test-clients/src/main/resource/contract/bytecodes/LargeContract.bin]: It's in `CryptoKitties.sol` . Will rename this accordingly.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/SelfDestructSuite.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
+package com.hedera.services.bdd.suites.contract.opcodes;
+
```
Comment (failfmi) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/opcodes/SelfDestructSuite.java:2]: nit: license
